### ISSUE TITLE
Background node cannot handle data-uris for textures (eg. backUrl) anymore!

### ISSUE THREAD

**Issue by elmkni** [x3dom/x3dom#971]
Background node cannot handle data-uris for textures (eg. backUrl) anymore!

In 1.7.2 everything worked! Could be used to insert textures on the fly with javascript/canvas2D.

 e.g.:

[v3_tt004_hwing.html.txt](https://github.com/x3dom/x3dom/files/3463828/v3_tt004_hwing.html.txt)

![v3_tt004_hwing html](https://user-images.githubusercontent.com/53652478/62412836-f7cdd980-b607-11e9-9032-db96d190d394.png)

**Comment by andreasplesch** [comment]
for experimentation:
https://jsfiddle.net/v5ughdjq/2/

**Comment by elmkni** [comment]
Hi Andreas,

thank you for your prompt answer. I tested already with a complete cubemap:

It works only with square images where width=height (even with scaling="true"),
otherwise webGL throws an error (at least in chrome 76) and only if all 6 images 
are set and not just the backUrl!

But that is not what I want! 

I want to be able to use 1 (one) background image, that statically occupies 
the complete background (streched if nessecary) as was the behavior with 
x3dom until v1.7.2 (which was a not-X3D-compliant feature, but very nice 
and usefull!).

Not a good idea to break a behavior that existed for ~10 years and upon 
which many people rely.

I tested alternatively to set style="background:url(data:....)" in \<X3D\>, which 
works, but makes it impossible to make a x3dom-snapshot including the back-
ground, where the model is rendered into the background image with nice 
antialiasing, which is the reason I used X3DOM (and not X3D) in the first place!

I also tested to set an \<ImageTexture\> on the fly: it works even with non-square
or partially transparent images! Why doesn't it work with \<Background\>?

Is it possible to restore the old behavior in cases where only the backUrl is set 
to a square/non-square image?

By the way: I also noticed, that I cannot copy/paste the debug log any more,
for example to get the viewpoint of a special positure of my model, to insert
it into the html-file.

I would like to use x3dom v1.8.0 because of its new additional features, like
NURBS etc. and because it uses webGL 2.0, but only if the background problem
and the debug-log in-accessibility can be fixed.

With best regards,

Elmar

**Comment by andreasplesch** [comment]
Thanks for the comprehensive report. I suspect the background changes have to do with supporting .dds in the backUrl field for PBR purposes. Unfortunately, I was not involved with these changes. Perhaps @tsturm has a quick fix on hand.

I also do not care for not being able to copy and paste from the debug log. But this can be fixed by css. The debug log uses the .x3dom-log class. Using user-select: text for its style should reenable selecting, for example by editing the default x3dom.css.

**Comment by elmkni** [comment]
Hi Andreas,

thank you for the 'user-select:text' advice -- it did the trick!

I used: \<style\> .x3dom-log{user-select:text} \</style\> in
my 'v3_tt004_hwing.html' file, so that 'x3dom.css' stays original.

By the way: the 'v'-switch still doesn't print the 'fieldOfView'-
value, which is actually important for reproducing the view!

About contacting Timo Sturm: I couldn't find his email address but if he
doesn't react to my bug report, he is probably the wrong man for fixing it.

I wonder who in the Fraunhofer Gesellschaft is responsible for bug-fixes.
Who decides which X3DOM-behavior is kept and which is changed without notice?
The \<Background\>-behavior change is not even mentioned in the release notes!

My (abstract) bug-fix proposal:

if ( ONLY 'backUrl' is set )
{
 old backdrop X3DOM v1.7.2 behavior
}
else { new cubemap X3DOM v1.8.0 behavior }

I hope it will be fixed soon...

With best regards,

Elmar

**Comment by andreasplesch** [comment]
That is good way to customize the css. x3dom relies exclusively on its community to help maintain and improve the code base. We are all grateful to FHG to host the web page, manage contributions and adapt x3dom for their own purposes.
Hacking x3dom is fun. You can learn and do a lot with just the debug version, the Chrome debugger and devtools' overrides.

**Comment by andreasplesch** [comment]
Could you please test with a regular image for Background backUrl ?

**Comment by elmkni** [comment]
Hi Andreas,

I have tested my scene with and without \<Background backUrl="test.jpg"\>
with firefox and chrome.

firefox didn't load the texture because of cors security errors!
chrome had absolutely no problem with a texture from my harddisk.

I appended my test files in: [test.zip](https://github.com/x3dom/x3dom/files/3466716/test.zip)

test.html
test.jpg
test.png
test.txt

With best regards,

Elmar

![test](https://user-images.githubusercontent.com/53652478/62451799-38943280-b76f-11e9-8c57-ff089fd50b3b.png)

the test results:
================================================================

with ENABLED \<Background\>

firefox 68 screen:
-------------------
X3DOM container-screen black

firefox 68 console:
-------------------
Array.map is deprecated; use Array.prototype.map instead x3dom.debug.js:14833:22
Use of Mutation Events is deprecated. Use MutationObserver instead. x3dom.debug.js:12774:21
Array.forEach is deprecated; use Array.prototype.forEach instead x3dom.debug.js:35852:23
Error: WebGL warning: drawElementsInstanced: One active vertex attrib (if any are active) must have a divisor of 0. 2 x3dom.debug.js:35174:16
Error: WebGL warning: texImage2D: Cross-origin elements require CORS. x3dom.debug.js:4007:6
SecurityError: The operation is insecure. x3dom.debug.js:4007
Error: WebGL warning: texImage2D: Cross-origin elements require CORS. x3dom.debug.js:4007:6
SecurityError: The operation is insecure. x3dom.debug.js:4007
Error: WebGL warning: drawElementsInstanced: One active vertex attrib (if any are active) must have a divisor of 0. 5 x3dom.debug.js:35174:16
Error: WebGL warning: readPixels: Out-of-bounds reads with readPixels are deprecated, and may be slow. x3dom.debug.js:32349:16
Error: WebGL warning: drawElementsInstanced: One active vertex attrib (if any are active) must have a divisor of 0. 18 x3dom.debug.js:35174:16

chrome 76 screen:
-----------------
X3DOM container-screen black

chrome 76 console:
------------------
no messages!

================================================================

with DISABLED \<Background\> (\<Switch\> does NOT disable, used XML-comments )

firefox 68 screen:
-------------------
X3DOM container-screen white (actually transparent) with BLACK spaceship

firefox 68 console:
-------------------
Array.map is deprecated; use Array.prototype.map instead x3dom.debug.js:14833:22
Use of Mutation Events is deprecated. Use MutationObserver instead. x3dom.debug.js:12774:21
Array.forEach is deprecated; use Array.prototype.forEach instead x3dom.debug.js:35852:23
Error: WebGL warning: texImage2D: Cross-origin elements require CORS. x3dom.debug.js:4007:6
SecurityError: The operation is insecure. x3dom.debug.js:4007
Error: WebGL warning: readPixels: Out-of-bounds reads with readPixels are deprecated, and may be slow. x3dom.debug.js:32349:16

chrome 76 screen:
-----------------
X3DOM container-screen white (actually transparent) with TEXTURED spaceship

chrome 76 console:
------------------
no messages!

**Comment by andreasplesch** [comment]
Thanks. I think the issue is unrelated to dataurls but a side effect of VR support.

fixed with #973 

You can test by building the background branch or by patching gfx_webgl.js as here in the codepen:

https://codepen.io/anon/pen/ZgXPVq?editors=1000

**Comment by andreasplesch** [comment]
As an aside, although this was in the end a small fix, it is generally advisable to use standard dom styling as much as possible, including for a static background. For a screenshot one could probably composite the background image and the webgl canvas together in a separate step.

In terms of the X3D standard, there is no guidance on how to deal with a single backUrl image, or any incomplete collection of cubemap images. There is a reference to allowing sky/groundColor together with images by not defining bottom/topUrls.

**Comment by elmkni** [comment]
Hi Andreas, 

I have just tested 'gfx_webgl.js'.

Simply included it after x3dom.js in my HTML file.

It works like magic -- all behavior like under X3DOM v1.7.2 ,
but with the additional cool new features of X3DOM v1.8.0!

I can even use non-square images as backgrounds, as I used to do!

Fantastic! -- If only I could understand how it works... ;)

About standard: X3DOM differs in many ways from pure X3D, and
its 'backUrl'-behavior was there for many years, which makes it a
de-facto standard!

About alternatively compositing: that's no option for me, because I use
X3DOM to be able to see my model in front of a static backdrop and
to be able to test different postures interactively (how they look in 
combination with the backdrop). 

There's no way to do it with pure X3D! Maybe a new node should be
introduced to X3D to make this possible, like '\<StaticBackdrop\>'!

By the way: combining a model-shot on transparent background with 
an image is not the same as rendering directly into the image/backbuffer, 
because the anti-aliasing doesn't take in account the colours of the 
backing image (the anti-aliasing is less good-looking)!

Believe me -- I have tested it: it looks different!

I really hope your fix (gfx_webgl.js) will be merged into the next X3DOM version!

Thanks a lot for your hard work!

With best regards,

Elmar

**Comment by andreasplesch** [comment]
The changes to gfx_webgl.js are very minor:
https://github.com/x3dom/x3dom/pull/973/commits/adb808fc9971e5fda45fa7cfb4cfb32fb8e83deb
It was a side effect of adding vr support but then overlooking this case.

Well, x3dom is not the only x3d browser. Most are not web based. I actually send out an inquiry on the x3d-public mailing list asking about x3d standard support for static image backdrops.

Once the PR in #973 is merged the dev version will include the flix. There are currently no plans for another release but typically after a major release there will be a number of such fixes which perhaps can then be collected in a 1.8.1 release.
